ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/truncate/TruncateTableAnalyzer.java

```diff
@@ -80,8 +86,25 @@ public void analyzeInternal(ASTNode root) throws SemanticException {
     checkTruncateEligibility(root, tableNode, tableNameString, table);
 
     Map<String, String> partitionSpec = getPartSpec((ASTNode) tableNode.getChild(1));
+
     addTruncateTableOutputs(tableNode, table, partitionSpec);
 
+    if (table.getStorageHandler() != null && !table.getStorageHandler().shouldTruncate(table, partitionSpec)) {
+      // Transform the truncate query to delete query.
+      StringBuilder rewrittenQuery = constructDeleteQuery(table, partitionSpec);
+      ReparseResult rr = ParseUtils.parseRewrittenQuery(conf, ctx, rewrittenQuery);
+      Context rewrittenCtx = rr.rewrittenCtx;
+      ASTNode rewrittenTree = rr.rewrittenTree;
+
+      BaseSemanticAnalyzer deleteAnalyzer = SemanticAnalyzerFactory.get(queryState, rewrittenTree);
```

Comment:
should we extend TruncateTableAnalyzer from RewriteSemanticAnalyzer?

Reply:
I think this is huge refactor since a lot of methods needs to be implemented. Its a separate task.

Reply:
I am ok if @kasakrisz doesn't have any concerns

Reply:
Actually extending `RewriteSemanticAnalyzer` means implementing two methods only:
*  `protected ASTNode getTargetTableNode(ASTNode tree)`
*  `protected void analyze(ASTNode tree, Table table, ASTNode tableName) throws SemanticException`

but the more problematic part is that currently `TruncateTableAnalyzer` extends `AbstractBaseAlterTableAnalyzer` and we can not extend from both classes.

Ideally the common parts of the rewrite logic should be implemented by a class which is not part of the `SemanticAnalyzer` class hierarchy hence that could be wrapped by SAs which needs rewrite. The use cases which needs rewrite seems to be increasing so sooner or later we should do such refactor.